### ISSUE TITLE
 Routed target `XXXX` does not exist, available targets are `[ds0, ds1]`.

### ISSUE THREAD

**Issue by cunzhiwang** [apache/shardingsphere#31661]
### Which version of ShardingSphere did you use?
```        
        <dependency>
            <groupId>org.apache.shardingsphere</groupId>
            <artifactId>shardingsphere-jdbc-core</artifactId>
            <version>5.4.1</version>
        </dependency>
```

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-JDBC

### Expected behavior
Successful routing and sharding in multithreaded situations

### Actual behavior
When I want to enhance datax(DataX is an open-source tool for offline synchronization of heterogeneous data sources from Alibaba), the following error occurs, but it is successful occasionally
``` java
org.apache.shardingsphere.sharding.exception.algorithm.sharding.ShardingRouteAlgorithmException: Routed target `ds0` does not exist, available targets are `[cmp_contract_0, cmp_contract_2, cmp_contract_4, cmp_contract_6]`.
	at org.apache.shardingsphere.sharding.route.strategy.type.standard.StandardShardingStrategy.doSharding(StandardShardingStrategy.java:80)
	at org.apache.shardingsphere.sharding.route.strategy.type.standard.StandardShardingStrategy.doSharding(StandardShardingStrategy.java:64)
	at org.apache.shardingsphere.sharding.route.engine.type.standard.ShardingStandardRoutingEngine.routeTables(ShardingStandardRoutingEngine.java:273)
	at org.apache.shardingsphere.sharding.route.engine.type.standard.ShardingStandardRoutingEngine.route0(ShardingStandardRoutingEngine.java:252)
	at org.apache.shardingsphere.sharding.route.engine.type.standard.ShardingStandardRoutingEngine.routeByShardingConditionsWithCondition(ShardingStandardRoutingEngine.java:142)
	at org.apache.shardingsphere.sharding.route.engine.type.standard.ShardingStandardRoutingEngine.routeByShardingConditions(ShardingStandardRoutingEngine.java:135)
	at org.apache.shardingsphere.sharding.route.engine.type.standard.ShardingStandardRoutingEngine.getDataNodes(ShardingStandardRoutingEngine.java:102)
	at org.apache.shardingsphere.sharding.route.engine.type.standard.ShardingStandardRoutingEngine.route(ShardingStandardRoutingEngine.java:84)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext0(ShardingSQLRouter.java:71)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:59)
	at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:47)
	at org.apache.shardingsphere.infra.route.engine.impl.PartialSQLRouteExecutor.route(PartialSQLRouteExecutor.java:69)
	at org.apache.shardingsphere.infra.route.engine.SQLRouteEngine.route(SQLRouteEngine.java:57)
	at org.apache.shardingsphere.infra.connection.kernel.KernelProcessor.route(KernelProcessor.java:60)
	at org.apache.shardingsphere.infra.connection.kernel.KernelProcessor.generateExecutionContext(KernelProcessor.java:51)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.createExecutionContext(ShardingSpherePreparedStatement.java:554)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.addBatch(ShardingSpherePreparedStatement.java:642)
```
![image](https://github.com/apache/shardingsphere/assets/57314096/0d985225-54a0-4325-ac58-c35c14dd9eaf)

### Reason analyze (If you can)
DataX is based on multi-threaded task execution. I think it is caused by thread synchronization issues.
### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
This is the code for getting the data source
```java
public static DataSource newInstance(File configDir) throws SQLException, ReflectiveOperationException, IOException {
        File yamlFile = ShardingConfigLoader.loadYamlFile(configDir);
        return YamlShardingSphereDataSourceFactory.createDataSource(yamlFile);
    }
```
This is the code for datax batch insert
```java
protected void doBatchInsert(Connection connection, List<Record> buffer)
                throws SQLException {
            PreparedStatement preparedStatement = null;
            try {
                connection.setAutoCommit(false);
                preparedStatement = connection
                        .prepareStatement(this.writeRecordSql);

                for (Record record : buffer) {
                    preparedStatement = fillPreparedStatement(
                            preparedStatement, record);
                    preparedStatement.addBatch();
                }
                preparedStatement.executeBatch();
                connection.commit();
            } catch (SQLException e) {
                LOG.warn("回滚此次写入, 采用每次写入一行方式提交. 因为:" + e.getMessage());
                connection.rollback();
                doOneInsert(connection, buffer);
            } catch (Exception e) {
                throw DataXException.asDataXException(
                        DBUtilErrorCode.WRITE_DATA_ERROR, e);
            } finally {
                DBUtil.closeDBResources(preparedStatement, null);
            }
        }
```
The corresponding implementation of PreparedStatement is ShardingSpherePreparedStatement. Here is a debug screenshot
![image](https://github.com/apache/shardingsphere/assets/57314096/27e8416b-79ad-4026-abfc-5212a631f133)
This is my config file
```
dataSources:
  ds0:
    dataSourceClassName: com.zaxxer.hikari.HikariDataSource
    driverClassName: com.mysql.cj.jdbc.Driver
    url: jdbc:mysql://127.0.0.1:3306/db_lab_orders_0?useSSL=false&useUnicode=true&characterEncoding=UTF-8&allowPublicKeyRetrieval=true
    username: root
    password: root
  ds1:
    dataSourceClassName: com.zaxxer.hikari.HikariDataSource
    driverClassName: com.mysql.cj.jdbc.Driver
    url: jdbc:mysql://127.0.0.1:3306/db_lab_orders_1?useSSL=false&useUnicode=true&characterEncoding=UTF-8&allowPublicKeyRetrieval=true
    username: root
    password: root

rules:
  - !SHARDING
    tables:
      cmp_contract:
        actualDataNodes: ds0.cmp_contract_$->{[0,2,4,6]},ds1.cmp_contract_$->{[1,3,5,7]}
     
        databaseStrategy:
          standard:
            shardingColumn: ID
            shardingAlgorithmName: database_inline
      
        tableStrategy:
          standard:
            shardingColumn: ID
            shardingAlgorithmName: table_inline
    
    
    shardingAlgorithms:
      database_inline:
        type: INLINE
        props:
          algorithm-expression: ds${ID % 2}
      table_inline:
        type: INLINE
        props:
          algorithm-expression: cmp_contract_${ID % 8}
```

**Comment by chaiduolai** [comment]
![image](https://github.com/apache/shardingsphere/assets/45263891/4f73aaff-2c1a-44b2-a048-3b7eabd2b604)
同样的错误

**Comment by cunzhiwang** [comment]
> ![image](https://private-user-images.githubusercontent.com/45263891/346887587-4f73aaff-2c1a-44b2-a048-3b7eabd2b604.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEyMDUxNDEsIm5iZiI6MTcyMTIwNDg0MSwicGF0aCI6Ii80NTI2Mzg5MS8zNDY4ODc1ODctNGY3M2FhZmYtMmMxYS00NGIyLWEwNDgtM2I3ZWFiZDJiNjA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE3VDA4MjcyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVhZjE0NjQ3MjBiYzkxNzY2NGQ1YTlmNTQzYzljYjUzNGNlMzI3MDNjYjQ3ODAyOWIzYTk3NWQyOWZhMmIyNTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MTZU-l_vprvmIinCXRZDEsJOCiP8Yb7HtZTHi-SEyFk) 同样的错误

You can use version 5.3.1, I can do it here

**Comment by jzx162050120** [comment]
Has this problem been solved？

**Comment by linghengqian** [comment]
- If you mean https://github.com/alibaba/DataX , what does your `enhance` do in `When I want to enhance datax(DataX is an open-source tool for offline synchronization of heterogeneous data sources from Alibaba), the following error occurs, but it is successful occasionally`? 

- One potential problem I see is that DataX does not have a docker image, which means it cannot be directly incorporated into shardingsphere's integration tests. 

- Anyone interested in this issue needs to provide a git with minimal reproducible unit tests.

**Comment by jzx162050120** [comment]
Not DataX,  the same error occurs when using the 5.4.1 version of Shardingsphere-jdbx-core. It is suspected that the project is not thread-safe when there is high concurrency.

**Comment by linghengqian** [comment]
- At least someone needs to provide reproducible operations. At least in the master branch, there is no known unit test that throws this error.

- If someone wants to reproduce using DataX, it seems easy for him to notice that the master branch of DataX has compilation failures.

**Comment by jzx162050120** [comment]
### Cause: java.sql.SQLException: Routed target `tableA_42` does not exist, available targets are `[tableB_32, tableB_33, tableB_34, tableB_35, tableB_36, tableB_37, tableB_38, tableB_39, tableB_40, tableB_41, tableB_42, tableB_43, tableB_44, tableB_45, tableB_46, tableB_47, tableB_48, tableB_49, tableB_50, tableB_51, tableB_52, tableB_53, tableB_54, tableB_55, tableB_56, tableB_57, tableB_58, tableB_59, tableB_60, tableB_61, tableB_62, tableB_71, tableB_99, tableB_101, tableB_102, tableB_103]`.
; uncategorized SQLException; SQL state [HY000]; error code [20081]; Routed target `tableA_42` does not exist, available targets are `[tableB_32, tableB_33, tableB_34, tableB_35, tableB_36, tableB_37, tableB_38, tableB_39, tableB_40, tableB_41, tableB_42, tableB_43, tableB_44, tableB_45, tableB_46, tableB_47, tableB_48, tableB_49, tableB_50, tableB_51, tableB_52, tableB_53, tableB_54, tableB_55, tableB_56, tableB_57, tableB_58, tableB_59, tableB_60, tableB_61, tableB_62, tableB_71, tableB_99, tableB_101, tableB_102, tableB_103]`.] with root cause
java.sql.SQLException: Routed target `tableA_42` does not exist, available targets are `[tableB_32, tableB_33, tableB_34, tableB_35, tableB_36, tableB_37, tableB_38, tableB_39, tableB_40, tableB_41, tableB_42, tableB_43, tableB_44, tableB_45, tableB_46, tableB_47, tableB_48, tableB_49, tableB_50, tableB_51, tableB_52, tableB_53, tableB_54, tableB_55, tableB_56, tableB_57, tableB_58, tableB_59, tableB_60, tableB_61, tableB_62, tableB_71, tableB_99, tableB_101, tableB_102, tableB_103]`.
	at org.apache.shardingsphere.infra.exception.core.external.sql.ShardingSphereSQLException.toSQLException(ShardingSphereSQLException.java:62)
	at org.apache.shardingsphere.infra.exception.dialect.SQLExceptionTransformEngine.toSQLException(SQLExceptionTransformEngine.java:54)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.execute(ShardingSpherePreparedStatement.java:425)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.ja...........

**Comment by linghengqian** [comment]
If you can reproduce it locally, you can directly open the PR and submit the unit test. I personally also have integration tests for a thousand real tables and did not see this problem.

**Comment by cunzhiwang** [comment]
Please wait a moment, I should be able to provide a reproducible example

**Comment by jzx162050120** [comment]
The prerequisite is high concurrent operations on two tables simultaneously.

**Comment by linghengqian** [comment]
> The prerequisite is high concurrent operations on two tables simultaneously.

It sounds like there is something wrong with your connection pool configuration. Or you are not using connection pool at all. You must provide unit tests.

**Comment by sunzheng04** [comment]
Has this problem been solved？The same error occurs when using the 5.4.1 version of Shardingsphere-jdbx-core. ShardingSpherePreparedStatement.execute() throws Exception.It is suspected that the project is not thread-safe when there is high concurrency.
`at org.apache.shardingsphere.infra.exception.core.external.sql.ShardingSphereSQLException.toSQLException(ShardingSphereSQLException.java:62)
 at org.apache.shardingsphere.infra.exception.dialect.SQLExceptionTransformEngine.toSQLException(SQLExceptionTransformEngine.java:54)
 at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.execute(ShardingSpherePreparedStatement.java:425)
 at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
 at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
 at jdk.internal.reflect.GeneratedMethodAccessor115.invoke(Unknown Source)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:568)
 at org.apache.ibatis.logging.jdbc.PreparedStatementLogger.invoke(PreparedStatementLogger.java:58)
 at jdk.proxy3/jdk.proxy3.$Proxy209.execute(Unknown Source)
 at org.apache.ibatis.executor.statement.PreparedStatementHandler.query(PreparedStatementHandler.java:65)
 at org.apache.ibatis.executor.statement.RoutingStatementHandler.query(RoutingStatementHandler.java:80)
 at org.apache.ibatis.executor.SimpleExecutor.doQuery(SimpleExecutor.java:65)
 at org.apache.ibatis.executor.BaseExecutor.queryFromDatabase(BaseExecutor.java:336)
 at org.apache.ibatis.executor.BaseExecutor.query(BaseExecutor.java:158)
 at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:110)
 at com.github.pagehelper.PageInterceptor.intercept(PageInterceptor.java:151)
 at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:59)
 at jdk.proxy2/jdk.proxy2.$Proxy207.query(Unknown Source)
 at org.apache.ibatis.session.defaults.DefaultSqlSession.selectL`

**Comment by linghengqian** [comment]
@sunzheng04 Please provide a stable and reproducible git repository with minimal reproducible unit tests in your issue. No one can read your mind.

**Comment by zshaojie** [comment]
We have the same issue.
**config**
rules:
  - !SHARDING
    tables:
      tb_user_draw_record:
        #        actualDataSources: db_0
        actualDataNodes: db_0.tb_user_draw_record_${0..9}
        tableStrategy:
          standard:
            shardingColumn: activity_id
            shardingAlgorithmName: tb_user_draw_record_hash_mod_10
      tb_user_draw_win_record:
        #        actualDataSources: db_0
        actualDataNodes: db_0.tb_user_draw_win_record_${0..9}
        tableStrategy:
          standard:
            shardingColumn: activity_id
            shardingAlgorithmName: tb_user_draw_win_record_hash_mod_10
    # 分片算法配置
    shardingAlgorithms:
      tb_user_draw_record_hash_mod_10:
        type: INLINE
        props:
          algorithm-expression: tb_user_draw_record_${activity_id % 10}
      tb_user_draw_win_record_hash_mod_10:
        type: INLINE
        props:
          algorithm-expression: tb_user_draw_win_record_${activity_id % 10}

**ERROR**
Caused by: java.sql.SQLException: Routed target `tb_user_draw_record_0` does not exist, available targets are `[tb_user_draw_win_record_0, tb_user_draw_win_record_1, tb_user_draw_win_record_2, tb_user_draw_win_record_3, tb_user_draw_win_record_4, tb_user_draw_win_record_5, tb_user_draw_win_record_6, tb_user_draw_win_record_7, tb_user_draw_win_record_8, tb_user_draw_win_record_9]`.
	at org.apache.shardingsphere.infra.exception.core.external.sql.ShardingSphereSQLException.toSQLException(ShardingSphereSQLException.java:62)
	at org.apache.shardingsphere.infra.exception.dialect.SQLExceptionTransformEngine.toSQLException(SQLExceptionTransformEngine.java:54)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.execute(ShardingSpherePreparedStatement.java:425)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
	at sun.reflect.GeneratedMethodAccessor336.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.ibatis.logging.jdbc.PreparedStatementLogger.invoke(PreparedStatementLogger.java:59)
	at com.sun.proxy.$Proxy197.execute(Unknown Source)
	at org.apache.ibatis.executor.statement.PreparedStatementHandler.query(PreparedStatementHandler.java:64)
	at org.apache.ibatis.executor.statement.RoutingStatementHandler.query(RoutingStatementHandler.java:79)
	at sun.reflect.GeneratedMethodAccessor335.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:64)
	at com.sun.proxy.$Proxy196.query(Unknown Source)
	at org.apache.ibatis.executor.SimpleExecutor.doQuery(SimpleExecutor.java:63)
	at org.apache.ibatis.executor.BaseExecutor.queryFromDatabase(BaseExecutor.java:325)
	at org.apache.ibatis.executor.BaseExecutor.query(BaseExecutor.java:156)

**Comment by linghengqian** [comment]
@zshaojie The master branch has a lot of similar unit tests that are not failing. You must extract the unit test that causes this error.

**Comment by interest2** [comment]
Hi all, the problem is about thread safety, and has actually been fixed by @FlyingZC in 2024.02，
In other words, the problem was fixed since 5.5.0, while it does exist below that version like 5.4.x.
![Image](https://github.com/user-attachments/assets/0d7d35d2-0e60-424d-81ba-a892e4c8d903)

The crucial change of that commit was removing @SingletonSPI from InlineExpressionParser.java
![Image](https://github.com/user-attachments/assets/1353f6e8-b02e-4b86-a0c7-77b7b6056fe1)

Why was it removed? It is related to the code below:
![Image](https://github.com/user-attachments/assets/ce82b935-2cbf-49a6-a3ea-a45119e85c0d)

When InlineExpressionParser is annotated with @SingletonSPI, the parser is shared across all threads, creating the possibility of data pollution between different threads.

Specifically, a valid SQL requires routing both the database index and the table index, and both go through the code below in InlineShardingAlgorithm.java:
![Image](https://github.com/user-attachments/assets/ad9f8f22-e937-4b53-810c-bea147b56dae)

**newInstance** method obtains an instance of GroovyInlineExpressionParser, and set the algorithmExpression of database or table;
**evaluateClosure** method retrieves groovy script from local cached map, using the expression above as the key.

In 5.4.x, that parser is shared (due to @SingletonSPI), so the algorithmExpression could be overwritten by another thread in high concurrency scenario. For example, this is what might happen in sequence:
1.   Thread A has just finished routing database index;
2.   Thread B sets the algorithmExpression of database for the parser; 
3.   Thread A overwrites it to the algorithmExpression of table;
4.   Thread B executes evaluateClosure(), will get the wrong groovy script because using the wrong algorithmExpression.

That’s why it hints: Routed target XXXX does not exist, because routed target is derived from the wrong routing script.

In summary, version since 5.5.0 fixed such thread safety issue by removing @SingletonSPI from InlineExpressionParser,  ensuring that each thread uses its own parser instance.
